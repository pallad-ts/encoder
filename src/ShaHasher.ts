import {webcrypto} from "node:crypto";
import {Hasher} from './Hasher';
import {TextBufferView} from "@pallad/text-buffer-view/compiled/TextBufferView";

export class ShaHasher extends Hasher {

	constructor(private variant: ShaHasher.Variant = 256) {
		super();
	}

	get algorithm() {
		return `SHA-${this.variant}`;
	}

	async hash(input: TextBufferView): Promise<TextBufferView> {
		const hashed = await webcrypto.subtle.digest(this.algorithm, input.originalArrayBuffer);
		return TextBufferView.fromArrayBuffer(hashed);
	}
}


export namespace ShaHasher {
	export type Variant = 256 | 384 | 512;
}
