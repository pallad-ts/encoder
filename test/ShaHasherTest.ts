import {ShaHasher} from "@src/ShaHasher";
import {TextBufferView} from "@pallad/text-buffer-view";

describe('ShaHasher', () => {
	describe.each<[ShaHasher.Variant]>([
		[256],
		[384],
		[512]
	])('variant %s', (variant) => {
		const hasher = new ShaHasher(variant);
		it('should hash a string', async () => {
			const input = 'd21bdb8e6f6f09a4a99541ab901bc1ed';
			const inputBuffer = TextBufferView.fromString(input, 'hex');

			const result = await hasher.hash(inputBuffer);
			expect(result.toString('hex')).toMatchSnapshot();
		});
	});
});
