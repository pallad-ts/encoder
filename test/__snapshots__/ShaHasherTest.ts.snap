// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShaHasher variant 256 should hash a string 1`] = `"7ac53faea4d4c95aefc9bbb85732ff3318d104cbfdce5dcb74090180e836fd1a"`;

exports[`ShaHasher variant 384 should hash a string 1`] = `"2af91beeee802ebace815f49d6441af678d989edaf2769798e3637d41da831f9d193205cde8024d0a8486f4a461b85a1"`;

exports[`ShaHasher variant 512 should hash a string 1`] = `"fbf8c005c45c0d36f88e861b42152e00a2f60171f0c62100799df018266d4fe2d7303dfde64f0b01a24206de2c684f22ed84767479818f49d0452b607ad3ecdf"`;
